control/cli.py

```diff
@@ -156,298 +207,1590 @@ def connect(self, host, port, client_key, client_cert, server_cert):
     ])
-    def resize_bdev(self, args):
-        """Resizes a bdev."""
-        req = pb2.resize_bdev_req(
-            bdev_name=args.bdev,
-            new_size=args.size,
-        )
-        ret = self.stub.resize_bdev(req)
-        self.logger.info(f"Resized bdev {args.bdev}: {ret.status}")
+    def version(self, args):
+        """Get CLI version"""
+        rc = 0
+        out_func, err_func = self.get_output_functions(args)
+        errmsg = ""
+        ver = os.getenv("NVMEOF_VERSION")
+        if not ver:
+            rc = errno.ENOKEY
+            errmsg = "Can't get CLI version"
+        else:
+            rc = 0
+            errmsg = os.strerror(0)
+        if args.format == "text" or args.format == "plain":
+            if not ver:
+                err_func(errmsg)
+            else:
+                out_func(f"CLI version: {ver}")
+        elif args.format == "json" or args.format == "yaml":
+            cli_ver = pb2.cli_version(status=rc, error_message=errmsg, version=ver)
+            out_ver = json_format.MessageToJson(cli_ver,
+                                                indent=4,
+                                                including_default_value_fields=True,
+                                                preserving_proto_field_name=True)
+            if args.format == "json":
+                out_func(f"{out_ver}")
+            elif args.format == "yaml":
+                obj = json.loads(out_ver)
+                out_func(yaml.dump(obj))
+        elif args.format == "python":
+            return pb2.cli_version(status=rc, error_message=errmsg, version=ver)
+        else:
+            assert False
+
+        return rc
+
+    def parse_version_string(self, version):
+        if not version:
+            return None
+        try:
+            vlist = version.split(".")
+            if len(vlist) != 3:
+                raise Exception
+            v1 = int(vlist[0])
+            v2 = int(vlist[1])
+            v3 = int(vlist[2])
+        except Exception:
+            return None
+        return (v1, v2, v3)
+
+    def gw_get_info(self):
+        ver = os.getenv("NVMEOF_VERSION")
+        req = pb2.get_gateway_info_req(cli_version=ver)
+        gw_info = self.stub.get_gateway_info(req)
+        if gw_info.status == 0:
+            base_ver = self.parse_version_string(BASE_GATEWAY_VERSION)
+            assert base_ver != None
+            gw_ver = self.parse_version_string(gw_info.version)
+            if gw_ver == None:
+                gw_info.status = errno.EINVAL
+                gw_info.bool_status = False
+                gw_info.error_message = f"Can't parse gateway version \"{gw_info.version}\"."
+            elif gw_ver < base_ver:
+                gw_info.status = errno.EINVAL
+                gw_info.bool_status = False
+                gw_info.error_message = f"Can't work with gateway version older than {BASE_GATEWAY_VERSION}"
```

Comment:
@epuertat there is also some code in grpc.py to return an error in case the CLI version is older than the gateway's one. But, we needed a way to give an error in case we use a gateway version older than 0.0.7 as the GRPC protocol is not compatible. As in such a case we can't rely on the server side, being older we need some code in the CLI side. I made sure that the "gw info" request remains compatible with the old versions so we could have that.

Reply:
Ok! 👍🏼